pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -136,6 +136,7 @@
 
     protected volatile MessageId lastDequeuedMessageId = MessageId.earliest;
     private volatile MessageId lastMessageIdInBroker = MessageId.earliest;
+    protected volatile CompletableFuture<Void> inProgressSubscribeFuture;
```

Comment:
I think it would be better to move this logic to `ConnectionHandler`. We should not protect `connectionOpen`, instead, we should protect `grabCnx`.

Reply:
I will open another PR for it. For the message lost issue specifically, this PR LGTM.

Reply:
I agree with you. The best way is to protect grabCnx, such as this:
- add a new [HandlerState](https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/HandlerState.java) `Registering(means sending "Subscribe" or "AddProducer" to broker)` between `Connecting` and `Reday`.
- `connectionOpened` is called only when comparing and swap(no reentrant) `Connecting -> Registering` successful

But if we did the change above, It would make a large impact. I suggest doing a fast fix first(and cherry-pick it into the previous branches), then doing the perfect changes above just in the newest release.